### PR TITLE
better help texts and sphinx reproduction of the help

### PR DESCRIPTION
**Pull request by enkore**
- Group options
- Nicer list of options in Sphinx

This is how the help looks like with the patch:

Sphinx: https://pstore.enkore.de/borg/usage.html#borg-create (this also works very nicely with RTFD's responsive theme, try making the browser window narrow :)

And on the console:

```
$ borg create --help                                   
usage: borg create [-h] [-v] [--debug] [--lock-wait N] [--show-version]
                   [--show-rc] [--no-files-cache] [--umask M]
                   [--remote-path PATH] [-n] [-s] [-p] [--list]
                   [--filter STATUSCHARS] [-e PATTERN]
                   [--exclude-from EXCLUDEFILE] [--exclude-caches]
                   [--exclude-if-present FILENAME] [--keep-tag-files] [-x]
                   [--numeric-owner] [--ignore-inode] [--read-special]
                   [--comment COMMENT] [--timestamp yyyy-mm-ddThh:mm:ss]
                   [-c SECONDS]
                   [--chunker-params CHUNK_MIN_EXP,CHUNK_MAX_EXP,HASH_MASK_BITS,HASH_WINDOW_SIZE]
                   [-C COMPRESSION]
                   ARCHIVE PATH [PATH ...]

Create new archive

positional arguments:
  ARCHIVE               name of archive to create (must be also a valid
                        directory name)
  PATH                  paths to archive

optional arguments:
  -h, --help            show this help message and exit
  -n, --dry-run         do not create a backup archive
  -s, --stats           print statistics for the created archive
  -p, --progress        show progress display while creating the archive,
                        showing Original, Compressed and Deduplicated sizes,
                        followed by the Number of files seen and the path
                        being processed, default: False
  --list                output verbose list of items (files, dirs, ...)
  --filter STATUSCHARS  only display items with the given status characters

Common options:
  -v, --verbose, --info
                        enable informative (verbose) output, work on log level
                        INFO
  --debug               enable debug output, work on log level DEBUG
  --lock-wait N         wait for the lock, but max. N seconds (default: 1).
  --show-version        show/log the borg version
  --show-rc             show/log the return code (rc)
  --no-files-cache      do not load/update the file metadata cache used to
                        detect unchanged files
  --umask M             set umask to M (local and remote, default: 0077)
  --remote-path PATH    set remote path to executable (default: "borg")

Exclusion options:
  -e PATTERN, --exclude PATTERN
                        exclude paths matching PATTERN
  --exclude-from EXCLUDEFILE
                        read exclude patterns from EXCLUDEFILE, one per line
  --exclude-caches      exclude directories that contain a CACHEDIR.TAG file
                        (http://www.brynosaurus.com/cachedir/spec.html)
  --exclude-if-present FILENAME
                        exclude directories that contain the specified file
  --keep-tag-files      keep tag files of excluded caches/directories

Filesystem options:
  -x, --one-file-system
                        stay in same file system, do not cross mount points
  --numeric-owner       only store numeric user and group identifiers
  --ignore-inode        ignore inode data in the file metadata cache used to
                        detect unchanged files.
  --read-special        open and read special files as if they were regular
                        files

Archive options:
  --comment COMMENT     add a comment text to the archive
  --timestamp yyyy-mm-ddThh:mm:ss
                        manually specify the archive creation date/time (UTC).
                        alternatively, give a reference file/directory.
  -c SECONDS, --checkpoint-interval SECONDS
                        write checkpoint every SECONDS seconds (Default: 300)
  --chunker-params CHUNK_MIN_EXP,CHUNK_MAX_EXP,HASH_MASK_BITS,HASH_WINDOW_SIZE
                        specify the chunker parameters. default: 19,23,21,4095
  -C COMPRESSION, --compression COMPRESSION
                        select compression algorithm (and level): none == no
                        compression (default), lz4 == lz4, zlib == zlib
                        (default level 6), zlib,0 .. zlib,9 == zlib (with
                        level 0..9), lzma == lzma (default level 6), lzma,0 ..
                        lzma,9 == lzma (with level 0..9).

This command creates a backup archive containing all files found while recursively
traversing all paths specified. The archive will consume almost no disk space for
files or parts of files that have already been stored in other archives.

To speed up pulling backups over sshfs and similar network file systems which do
not provide correct inode information the --ignore-inode flag can be used. This
potentially decreases reliability of change detection, while avoiding always reading
all files on these file systems.

See the output of the "borg help patterns" command for more help on exclude patterns.
```

The large unhelpful chunk of options at the top of the console output is generated by some code buried in implementation details, so I didn't change that. The code in setup.py uses implementation details as well, but only superficial ones (_actions, _action_groups).

Note that these argument groups are purely cosmetic and don't change the semantics of the command line.

### CODE REVIEW

**Comment by enkore** [comment]
Hm, I guess we could put this into 1.0-maint, it's a docs change, not a functionality change; the [argparse](https://docs.python.org/3/library/argparse.html#argument-groups) docs guarantee that moving options to groups does not change behaviour of the parser except for help display.

**Comment by ThomasWaldmann** [comment]
please don't merge anything into 1.0-maint now, i am currently working on releasing 1.0.1.

File: docs/usage.rst [hunk 1]
```diff
@@ -179,6 +179,12 @@ For more information about that, see: https://xkcd.com/1179/
 Unless otherwise noted, we display local date and time.
 Internally, we store and process date and time as UTC.
 
+Common options
+~~~~~~~~~~~~~~
+
+All Borg commands share these options:
+
+.. include:: usage/--common.rst.inc
```
Comment (ThomasWaldmann) [inline docs/usage.rst]: please don't generate filenames that start with --.

how about options-common.rst.inc?

Comment (enkore) [inline docs/usage.rst]: Ah yes, the idea was to have something that can't ever clash with a command name, but `--something` files are also notoriously annoying to handle on the command line. Changed it ;)

File: docs/borg_theme/css/borg.css [hunk 2]
```diff
@@ -16,3 +16,7 @@
 .wy-side-nav-search > div.version {
     color: rgba(255, 255, 255, 0.5);
 }
+
+#usage dt code {
+    font-weight: normal;
+}
```
Comment (ThomasWaldmann) [inline docs/borg_theme/css/borg.css]: line end missing?

File: docs/usage.rst [hunk 3]
```diff
@@ -191,6 +191,12 @@ For more information about that, see: https://xkcd.com/1179/
 Unless otherwise noted, we display local date and time.
 Internally, we store and process date and time as UTC.
 
+Common options
+~~~~~~~~~~~~~~
+
+All Borg commands share these options:
```
Comment (ThomasWaldmann) [inline docs/usage.rst]: All |project_name| commands ...

File: setup.py [hunk 4]
```diff
@@ -168,12 +170,72 @@ def run(self):
...
+                option_fmt = '``%%s %s``' % option.metavar
+            else:
+                option_fmt = '``%s``'
+            option_str = ', '.join(option_fmt % s for s in option.option_strings)
+            options.append((option_str, option))
+        for option_str, option in options:
+            help = textwrap.indent(get_help(option), ' ' * 4)
+            text.append(option_str)
+            text.append(help)
+        shipout(text)
```
Comment (ThomasWaldmann) [inline setup.py]: you access a lot of ._xxx attributes in above code - shall we rely on their (undocumented?) semantics?

Comment (enkore) [inline setup.py]: They are implementation details, but they're pretty central to how argparse does it's stuff, so I deemed them unlikely to change. I wouldn't put this in the production code, though.

**Pull request merged**